src/internal/storage/chunk/uploader.go

```diff
@@ -76,6 +81,97 @@ func (u *Uploader) Upload(meta interface{}, r io.Reader) error {
 }
 
 func (u *Uploader) Copy(meta interface{}, dataRefs []*DataRef) error {
+	var stableDataRefs, nextDataRefs []*DataRef
+	taskChain := taskchain.New(u.ctx, u.chunkSem)
+	appendDataRefs := func(dataRefs []*DataRef) error {
+		return taskChain.CreateTask(func(_ context.Context) (func() error, error) {
+			return func() error {
+				for _, dataRef := range dataRefs {
+					stableDataRefs = append(stableDataRefs, proto.Clone(dataRef).(*DataRef))
+				}
+				return nil
+			}, nil
+		})
+	}
+	for i := 0; i < len(dataRefs); i++ {
+		if !stableDataRef(dataRefs[i]) {
+			if err := appendDataRefs(nextDataRefs); err != nil {
+				return err
+			}
+			nextDataRefs = nil
+			var err error
+			i, err = u.align(u.ctx, dataRefs, i, func(chunk []byte) error {
+				return taskChain.CreateTask(func(ctx context.Context) (func() error, error) {
+					dataRef, err := upload(ctx, u.client, chunk, nil, u.noUpload)
+					if err != nil {
+						return nil, err
+					}
+					return func() error {
+						stableDataRefs = append(stableDataRefs, dataRef)
+						return nil
+					}, nil
+				})
+			})
+			if err != nil {
+				return err
+			}
+			continue
+		}
+		nextDataRefs = append(nextDataRefs, dataRefs[i])
+	}
+	if err := appendDataRefs(nextDataRefs); err != nil {
+		return err
+	}
+	return u.taskChain.CreateTask(func(_ context.Context) (func() error, error) {
+		if err := taskChain.Wait(); err != nil {
+			return nil, err
+		}
+		if len(dataRefs) > 0 {
```

Comment:
This refers to dataRefs potentially after Copy has returned.  I think that's fine because the slice header (including the length) is passed by value.  Accessing this would not be okay. 

I think a comment on the function explaining how dataRefs is or isn't retained would be helpful.

Reply:
This has been updated to use stableDataRefs, which it should have been using anyways.